Add APDU debugging using a trace api similar to httptrace

areese opened the pull request:
@ericchiang Please let me know what you think of this way of tracing.
I need to sign the commits, I don't have the keys right now, and I need to try this out, I don't have the harness.

ericchiang reviewed (commented):
Rather than using contexts, I would recommend the exported API look something:

```go
type ClientTrace struct {
    Transmit func(req []byte)
    TransmitResult func(...)
}
type Client struct {
    Trace *ClientTrace
}
func (c *Client) Open(*Yubikey, error) { ... }
```

Dropping the WithClientTrace methods and use of contexts

areese commented:
> Rather than using contexts, I would recommend the exported API look something:
> 
> ```go
> type ClientTrace struct {
>     Transmit func(req []byte)
>     TransmitResult func(...)
> }
> type Client struct {
>     Trace *ClientTrace
> }
> func (c *Client) Open(*Yubikey, error) { ... }
> ```
> 
> Dropping the WithClientTrace methods and use of contexts

I think that's way better.  I didn't think of that path, and just added context.

areese committed:
Add APDU debugging using a trace api similar to httptrace

A ClientTrace struct can be set on a YubiKey object,
or a piv.Client struct can be created and Open called against that.
